File: .claude/skills/sast-scan/rules/semgrep/go/tests/test_taint_rules.go
package main

import (
	"database/sql"
	"fmt"
	"html/template"
	"io"
	"net/http"
	"os"
	"os/exec"
)

// --- taint-sql-injection ---

func searchHandler_Unsafe(w http.ResponseWriter, r *http.Request) {
	q := r.FormValue("q")
	db, _ := sql.Open("postgres", "...")
	// ruleid: go.security.taint-sql-injection
	db.Query("SELECT * FROM users WHERE name LIKE '%" + q + "%'")
}

func searchHandler_UnsafeFmt(w http.ResponseWriter, r *http.Request) {
	q := r.URL.Query().Get("q")
	sqlStr := fmt.Sprintf("SELECT * FROM users WHERE name='%s'", q)
	db, _ := sql.Open("postgres", "...")
	// ruleid: go.security.taint-sql-injection
	db.Exec(sqlStr)
}

func searchHandler_Safe(w http.ResponseWriter, r *http.Request) {
	q := r.FormValue("q")
	db, _ := sql.Open("postgres", "...")
	// ok: go.security.taint-sql-injection
	db.Query("SELECT * FROM users WHERE name LIKE $1", "%"+q+"%")
}

// --- taint-command-injection ---

func pingHandler_Unsafe(w http.ResponseWriter, r *http.Request) {
	host := r.FormValue("host")
	// ruleid: go.security.taint-command-injection
	exec.Command("sh", "-c", "ping -c 1 "+host).Run()
}

func pingHandler_UnsafeFmt(w http.ResponseWriter, r *http.Request) {
	host := r.URL.Query().Get("host")
	cmdStr := fmt.Sprintf("ping -c 1 %s", host)
	// ruleid: go.security.taint-command-injection
	exec.Command("sh", "-c", cmdStr).Run()
}

func pingHandler_Safe(w http.ResponseWriter, r *http.Request) {
	// ok: go.security.taint-command-injection
	exec.Command("ping", "-c", "1", "localhost").Run()
}

// --- taint-path-traversal ---

func exportHandler_Unsafe(w http.ResponseWriter, r *http.Request) {
	filename := r.FormValue("file")
	path := "/var/exports/" + filename
	// ruleid: go.security.taint-path-traversal
	os.Open(path)
}

func exportHandler_Safe(w http.ResponseWriter, r *http.Request) {
	filename := r.FormValue("file")
	clean := filepath.Clean(filename)
	// ok: go.security.taint-path-traversal
	os.Open("/var/exports/" + filepath.Base(clean))
}

// --- taint-ssrf ---

func proxyHandler_Unsafe(w http.ResponseWriter, r *http.Request) {
	url := r.FormValue("url")
	// ruleid: go.security.taint-ssrf
	http.Get(url)
}

func proxyHandler_UnsafePost(w http.ResponseWriter, r *http.Request) {
	endpoint := r.URL.Query().Get("callback")
	// ruleid: go.security.taint-ssrf
	http.Post(endpoint, "application/json", nil)
}

func proxyHandler_Safe(w http.ResponseWriter, r *http.Request) {
	// ok: go.security.taint-ssrf
	http.Get("https://api.example.com/health")
}

// --- taint-xss-template ---

func greetHandler_Unsafe(w http.ResponseWriter, r *http.Request) {
	name := r.FormValue("name")
	tmpl, _ := template.New("greet").Parse("Hello {{.}}")
	// ruleid: go.security.taint-xss-template
	tmpl.Execute(w, name)
}

func greetHandler_UnsafeFprintf(w http.ResponseWriter, r *http.Request) {
	name := r.URL.Query().Get("name")
	// ruleid: go.security.taint-xss-template
	fmt.Fprintf(w, "<h1>Hello %s</h1>", name)
}

func greetHandler_Safe(w http.ResponseWriter, r *http.Request) {
	name := r.FormValue("name")
	// ok: go.security.taint-xss-template
	io.WriteString(w, template.HTMLEscapeString(name))
}

// --- taint-open-redirect ---

func redirectHandler_Unsafe(w http.ResponseWriter, r *http.Request) {
	target := r.FormValue("redirect")
	// ruleid: go.security.taint-open-redirect
	http.Redirect(w, r, target, 302)
}

func redirectHandler_Safe(w http.ResponseWriter, r *http.Request) {
	// ok: go.security.taint-open-redirect
	http.Redirect(w, r, "/home", 302)
}

File: .claude/skills/sast-scan/rules/semgrep/ruby/tests/test_taint_rules.rb
# Ruby taint-mode rule test fixtures.
# Verifies that Semgrep tracks data flow from params through assignments to sinks.

class TaintSecurityController < ApplicationController
  # --- taint-sql-injection ---

  def search_unsafe
    q = params[:q]
    # ruleid: ruby.security.taint-sql-injection
    User.where("name LIKE '%#{q}%'")
  end

  def search_unsafe_find_by_sql
    keyword = params.fetch(:keyword)
    # ruleid: ruby.security.taint-sql-injection
    User.find_by_sql("SELECT * FROM users WHERE name LIKE '%#{keyword}%'")
  end

  def search_unsafe_execute
    name = params["name"]
    sql = "SELECT * FROM users WHERE name = '#{name}'"
    # ruleid: ruby.security.taint-sql-injection
    ActiveRecord::Base.connection.execute(sql)
  end

  def search_safe
    q = params[:q]
    # ok: ruby.security.taint-sql-injection
    User.where("name LIKE ?", "%#{q}%")
  end

  # --- taint-command-injection ---

  def ping_unsafe
    host = params[:host]
    # ruleid: ruby.security.taint-command-injection
    system("ping -c 1 #{host}")
  end

  def run_unsafe_backticks
    script = params[:script]
    # ruleid: ruby.security.taint-command-injection
    `sh /tmp/#{script}`
  end

  def ping_safe
    host = params[:host].to_i
    # ok: ruby.security.taint-command-injection
    system("ping", "-c", "1", "host-#{host}")
  end

  # --- taint-path-traversal ---

  def export_unsafe
    filename = params[:file]
    path = Rails.root.join("exports", filename)
    # ruleid: ruby.security.taint-path-traversal
    send_file path
  end

  def read_unsafe
    name = params[:name]
    # ruleid: ruby.security.taint-path-traversal
    File.read("/data/#{name}.json")
  end

  def export_safe
    filename = File.basename(params[:file])
    # ok: ruby.security.taint-path-traversal
    send_file Rails.root.join("exports", filename)
  end

  # --- taint-ssrf ---

  def proxy_unsafe
    url = params[:url]
    # ruleid: ruby.security.taint-ssrf
    Net::HTTP.get(URI.parse(url))
  end

  def fetch_unsafe
    endpoint = params[:callback]
    # ruleid: ruby.security.taint-ssrf
    RestClient.get(endpoint)
  end

  def fetch_safe
    # ok: ruby.security.taint-ssrf
    Net::HTTP.get(URI.parse("https://api.example.com/health"))
  end

  # --- taint-xss ---

  def greet_unsafe
    name = params[:name]
    greeting = "Hello, " + name
    # ruleid: ruby.security.taint-xss
    render inline: "<h1>#{greeting}</h1>"
  end

  def raw_output_unsafe
    msg = params[:message]
    # ruleid: ruby.security.taint-xss
    raw(msg)
  end

  def greet_safe
    name = params[:name]
    # ok: ruby.security.taint-xss
    render plain: h(name)
  end

  # --- taint-deserialize ---

  def import_unsafe
    data = params[:payload]
    # ruleid: ruby.security.taint-deserialize
    YAML.load(data)
  end

  def load_unsafe_marshal
    blob = cookies[:session]
    raw = Base64.decode64(blob)
    # ruleid: ruby.security.taint-deserialize
    Marshal.load(raw)
  end

  def import_safe
    data = params[:config]
    # ok: ruby.security.taint-deserialize
    YAML.safe_load(data)
  end
end
